clustering/server/src/main/java/org/wildfly/clustering/server/singleton/ServiceLifecycle.java

```diff
@@ -57,36 +66,54 @@ public ServiceLifecycle(ServiceController<?> controller) {
 
     @Override
     public void start() {
-        this.transition(State.UP);
+        this.transition(Transition.START);
     }
 
     @Override
     public void stop() {
-        this.transition(State.DOWN);
+        this.transition(Transition.STOP);
     }
 
-    private void transition(State targetState) {
+    private void transition(Transition transition) {
         // Short-circuit if the service is already at the target state
-        if (this.controller.getState() == targetState) return;
+        if (this.isComplete(transition)) return;
+
+        CountDownLatch latch = new CountDownLatch(1);
+        LifecycleListener listener = new CountDownLifecycleListener(latch, transition.targetEvents);
+        this.controller.addListener(listener);
 
-        StabilityMonitor monitor = new StabilityMonitor();
-        monitor.addController(this.controller);
         try {
+            if (this.isComplete(transition)) return;
+
             // Force service to transition to desired state
-            Mode targetMode = modeToggle.get(targetState).get(this.controller.getMode());
-            if (targetMode != null) {
-                this.controller.setMode(targetMode);
+            Mode currentMode = this.controller.getMode();
+            Mode targetMode = transition.modeTransitions.get(currentMode);
+            if (targetMode == null) {
+                throw new IllegalStateException(currentMode.name());
             }
 
-            monitor.awaitStability();
+            this.controller.setMode(targetMode);
+
+            latch.await();
```

Comment:
This latch.await() has the same concurrency issues as had code code before. See WFLY-12601, issue (2) description.

Reply:
Can you elaborate?  Your comment in WFLY-12601 seems specific obtaining service values, but ServiceLifecycle does not read service values - it merely seeks to start or stop a given service by transitioning it between Mode.NEVER and Mode.ACTIVE.